### ISSUE TITLE
Use other model file based on Flux-schnell?

### ISSUE THREAD

**Issue by puppyapple** [filipstrand/mflux#102]
Thanks for the great work!
Recently I see new models like [shuttleai/shuttle-3-diffusion · Hugging Face](https://huggingface.co/shuttleai/shuttle-3-diffusion), which is finetuned on flux-schnell. 
Is there any way to use mflux on these models?

**Comment by anthonywu** [comment]
@puppyapple if you're up for installing a dev build, try:

```sh
pip install git+https://github.com/anthonywu/mflux.git@support-hf-models

mflux-generate \
  --base-model schnell \
  --steps 4 \
  --model shuttleai/shuttle-3-diffusion \
  --prompt "A dog holding a sign that says mflux rocks"
```

make sure to accept the repo terms of service first at: https://huggingface.co/shuttleai/shuttle-3-diffusion before trying to use mflux / huggingface-hub to download the models to disk

sample from my local gen using the branch in #103 

![image](https://github.com/user-attachments/assets/8be880da-646a-4a25-8236-0326b104ab12)

**Comment by tariks** [comment]
@anthonywu for this syntax, how does quant work? for shuttle-3-diffusion, what's the difference between --quant 8 on full model or cloning the fp8 version?

**Comment by anthonywu** [comment]
Here's my investigation.

## TL;DR

- The mflux-converted q8 model works as I confirm in this reply.
- the upstream `fp8` doesn't work because that HF repo does not contain the full assets. You may be able to get it to work by locally stitching the repo assets together.

## download the official models

```sh
# official bfloat16 model
huggingface-cli download shuttleai/shuttle-3-diffusion 

# official fp8 model
huggingface-cli download shuttleai/shuttle-3-diffusion-fp8
```

### observe the model size on disk

```sh
du -sh ~/.cache/huggingface/hub/models--shuttleai--shuttle-3-diffusion*
 54G	~/.cache/huggingface/hub/models--shuttleai--shuttle-3-diffusion
 11G	~/.cache/huggingface/hub/models--shuttleai--shuttle-3-diffusion-fp8
```

## use mflux to save a q8 model

this should work in the latest commit of PR #103

```sh
mflux-save \
  -m shuttleai/shuttle-3-diffusion \
  --base-model schnell \
  -q 8 \
  --path /tmp/shuttle-3-diffusion-q8
```

this produces a local 17G converted model (compared to the 11G fp8 produced by their util: https://huggingface.co/shuttleai/shuttle-3-diffusion-fp8/blob/main/convert.py)

```sh
du -sh /tmp/shuttle-3-diffusion-q8
 17G	/tmp/shuttle-3-diffusion-q8
```

In mflux's `model_saver.py`, the conversion is passed to `mlx.core. save_safetensors`, this is its `help()`:

```
save_safetensors = <nanobind.nb_func object>
    save_safetensors(file: str, arrays: dict[str, array], metadata: Optional[dict[str, str]] = None)

    Save array(s) to a binary file in ``.safetensors`` format.

    See the `Safetensors documentation
    <https://huggingface.co/docs/safetensors/index>`_ for more
    information on the format.

    Args:
        file (file, str): File in which the array is saved.
        arrays (dict(str, array)): The dictionary of names to arrays to
        be saved. metadata (dict(str, str), optional): The dictionary of
        metadata to be saved.
```

the caller is:

```py
    def save_weights(base_path: str, bits: int, model: nn.Module, subdir: str):
        path = Path(base_path) / subdir
        path.mkdir(parents=True, exist_ok=True)
        weights = ModelSaver._split_weights(base_path, dict(tree_flatten(model.parameters())))
        for i, weight in enumerate(weights):
            mx.save_safetensors(
                str(path / f"{i}.safetensors"),
                weight,
                {"quantization_level": str(bits)},
            )
```

so the difference between the official fp8 17G version and the mflux converted 11G version is a matter of `mlx` vs `PyTorch` low level details of the quantization

## compare outputs

### generating with official bfloat16

```sh
mflux-generate \
  --seed 42 \
  --base-model schnell \
  --model shuttleai/shuttle-3-diffusion \
  --prompt "a cat holding a sign saying meow" \
  --steps 4
```

![image](https://github.com/user-attachments/assets/ff0ec243-b4ba-4bf2-9889-828fc195ba1c)

### generating with mflux-converted q8

```sh
mflux-generate \
  --seed 42 \
  --model schnell \
  --path /tmp/shuttle-3-diffusion-q8 \
  --prompt "a cat holding a sign saying meow" \
  --steps 4 
```

![image](https://github.com/user-attachments/assets/9e3d1155-ce36-4a71-a9a8-4fa5663358f6)

the two images look same-ish to me, but `diff` of the two files reports a difference.

### generating with upstream fp8

```sh
mflux-generate \
  --seed 42 \
  --base-model schnell \
  --model shuttleai/shuttle-3-diffusion-fp8 \
  --prompt "a cat holding a sign saying meow" \
  --steps 4
```

this doesn't actually work out of the box because the HF repo for fp8 does not contain all the assets from the bfloat16 version! And that's probably why it weighs in a 11G rather than closer to 17G, so a full repo may not be much diff in size after all.

<img width="717" alt="image" src="https://github.com/user-attachments/assets/82321571-d47c-4d43-889f-c566671459fe">

the error from mflux is `OSError: Incorrect path_or_model_id: '/Users/anthonywu/.cache/huggingface/hub/models--shuttleai--shuttle-3-diffusion-fp8/snapshots/c91030ae2199a2a57f9e0814f6f427692f15aabd/tokenizer'. Please provide either the path to a local folder or the repo_id of a model on the Hub.` and the reason is the `tokenizer` and adjacent assets are not in the fp8 repo.

if the official fp8 repo is laid out like the full version then I think the HF model would *just work*

<img width="681" alt="image" src="https://github.com/user-attachments/assets/ab62ed9e-698a-4125-8ca6-eb234da9b4b7">

so as of this writing - the only way to get out of box q8 functionality is to use mflux to convert to q8 locally

this does create a todo to better display the errors for HF repos that do not present its full assets

**Comment by tariks** [comment]
Thank you! Happy to confirm all this works as you say with the pr

**Comment by puppyapple** [comment]
> @puppyapple if you're up for installing a dev build, try:
> 
> ```shell
> pip install git+https://github.com/anthonywu/mflux.git@support-hf-models
> 
> mflux-generate \
>   --base-model schnell \
>   --steps 4 \
>   --model shuttleai/shuttle-3-diffusion \
>   --prompt "A dog holding a sign that says mflux rocks"
> ```
> 
> make sure to accept the repo terms of service first at: https://huggingface.co/shuttleai/shuttle-3-diffusion before trying to use mflux / huggingface-hub to download the models to disk
> 
> sample from my local gen using the branch in #103
> 
> ![image](https://private-user-images.githubusercontent.com/462072/389135664-8be880da-646a-4a25-8236-0326b104ab12.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI3NzExNDcsIm5iZiI6MTczMjc3MDg0NywicGF0aCI6Ii80NjIwNzIvMzg5MTM1NjY0LThiZTg4MGRhLTY0NmEtNGEyNS04MjM2LTAzMjZiMTA0YWIxMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOFQwNTE0MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MzAxZDA3ZTFmNzQwYjljY2ZiZWFhNTkyM2ZmNmY1ZWRhMmQwNmM5NDk0Mzg3MmMxNTc2OTkxNWYyMTZjNjQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.OJvYwwVIrpfZXrP0FxDAmBgn7hYxTYZfoPCI-imDNAU)

thanks so much! that works perfectly for me！

**Comment by anthonywu** [comment]
@puppyapple the PR hasn't officially merged yet, don't forget to circle back and re-install `pip install mflux` when the new version publishes.
